ws/config.py

```diff
@@ -1,180 +1,232 @@
+    
+
+class _ActionCheckDirname(argparse.Action):
+
+    def __call__(self, parser, namespace, path, option_string=None):
+        dirname = os.path.dirname(path)
+        if not os.path.isdir(dirname)
+            raise FileNotFoundError(
+                "no such directory: {}".format(dirname))
+        setattr(namespace, self.dest, path)
+
+
+def getArgParser(**kwargs):
+    """
+    Create an instance of :py:class:`_ArgumentParser`. Make its initial setup
+    and set the logging arguments.
+
+    :param kwargs: passed to :py:class:`_ArgumentParser()` constructor.
+    :returns: an instance of :py:class:`_ArgumentParser`.
+    """
+    ap = _ArgumentParser(**kwargs)
+    ap.setup()
 
     # include logging arguments into the global group
     ws.logging.set_argparser(ap)
 
-    # add other global arguments
-    ap.add_argument("--cache-dir", type=argtype_dirname_must_exist, metavar="PATH",
-            help="directory for storing cached data (will be created if necessary, but parent directory must exist) (default: %(default)s)")
-
-    ap.set_defaults(**Defaults())
-
     return ap
 
-def object_from_argparser(klass, subname=None, *args, **kwargs):
+
+def object_from_argparser(klass, section=None, **kwargs):
     """
     Create an instance of ``klass`` using its :py:meth:`klass.from_argparser()`
-    factory and a clean instance of :py:class:`argparse.ArgumentParser`. On top
-    of that, logging interface is set up using the :py:mod:`ws.logging` module.
+    factory and a instance of :py:class:`_ArgumentParser`. On top of that, 
+    logging interface is set up using the :py:mod:`ws.logging` module.
 
     :param klass: the class to instantiate
-    :param subname:
+    :param section:
         The name of the subsection to be read from the configuration file
         (usually the name of the script). By default ``sys.argv[0]`` is taken.
-    :param args: passed to :py:class:`argparse.ArgumentParser()` constructor
-    :param kwargs: passed to :py:class:`argparse.ArgumentParser()` constructor
+    :param kwargs: passed to :py:class:`_ArgumentParser()` constructor
     :returns: an instance of :py:class:`klass`
     """
-    argparser = getArgParser(*args, **kwargs)
+
+    # argparser creation, initial setup and parsing sys.argv
+    # for config/cache-dir/etc options
+    argparser = getArgParser(**kwargs)
+    argparser.setup()
+    args, remaining_argv = argparser.parse_known_args()
+    
+    # read the config file and fetch the script-related section options
+    cfp = ConfigParser(args.config)
+    config_args = cfp.fetch_section(section)
+    
+    # class parser setup and final parsing
     klass.set_argparser(argparser)
-    args = argparser.parse_args()
+    argparser.parse_args(config_args + remaining_argv, namespace=args)
 
     # set up logging
     ws.logging.init(args)
-
-    # TODO: depends on ConfigArgParse, in case of argparse just log the Namespace
     logger.debug("Parsed arguments:\n" + argparser.format_values())
```

Comment:
Please replace this with `logger.debug("Parsed arguments:\n{}".format(args))` (that's what the TODO note meant).

Reply:
Already done (it was throwing an error).